### PR TITLE
Migrate Buffer to Kotlin

### PR DESCRIPTION
**Pull request by swankjesse**
This first pass is intended to get things converted over and correct. We need to take
another pass in follow-up to replace overloaded methods with JvmOverloads and default
parameters. Also to consider using vals where we currently use funs, and other
Kotlin niceties.

### CODE REVIEW

File: okio/jvm/src/main/java/okio/Buffer.kt [hunk 1]
```diff
@@ -0,0 +1,2176 @@
...
+
+      val data = segment!!.data
+      var pos = segment.pos
+      val limit = segment.limit
+
+      while (pos < limit) {
+        val digit: Int
+
+        val b = data[pos]
+        if (b >= '0'.toByte() && b <= '9'.toByte()) {
```
Comment (swankjesse) [inline okio/jvm/src/main/java/okio/Buffer.kt]: for follow-up, these should be ranges

File: okio/jvm/src/main/java/okio/Buffer.kt [hunk 2]
```diff
@@ -0,0 +1,2176 @@
...
+  }
+
+  override fun writeHexadecimalUnsignedLong(v: Long): Buffer {
+    var v = v
+    if (v == 0L) {
+      // Both a shortcut and required since the following code can't handle zero.
+      return writeByte('0'.toInt())
+    }
+
+    val width = java.lang.Long.numberOfTrailingZeros(java.lang.Long.highestOneBit(v)) / 4 + 1
```
Comment (swankjesse) [inline okio/jvm/src/main/java/okio/Buffer.kt]: Obvious candidate for platform.kt

File: okio/jvm/src/main/java/okio/Buffer.kt [hunk 3]
```diff
@@ -0,0 +1,2176 @@
...
+  /** Returns the 128-bit MD5 hash of this buffer.  */
+  @Suppress("NON_FINAL_MEMBER_IN_FINAL_CLASS") // Required to keep JApicmp happy.
+  open fun md5() = digest("MD5")
+
+  /** Returns the 160-bit SHA-1 hash of this buffer.  */
+  @Suppress("NON_FINAL_MEMBER_IN_FINAL_CLASS") // Required to keep JApicmp happy.
+  open fun sha1() = digest("SHA-1")
+
+  /** Returns the 256-bit SHA-256 hash of this buffer.  */
+  @Suppress("NON_FINAL_MEMBER_IN_FINAL_CLASS") // Required to keep JApicmp happy.
```
Comment (swankjesse) [inline okio/jvm/src/main/java/okio/Buffer.kt]: I’m tempted to cut an intermediate 1.x release that makes all of these final, just so we have that as our baseline

Comment (JakeWharton) [reply okio/jvm/src/main/java/okio/Buffer.kt]: Alternatively, once we're ready for 2.0 we disable japicmp, fix all these, release, and make 2.0 the baseline.

Comment (brettwooldridge) [reply okio/jvm/src/main/java/okio/Buffer.kt]: Just a fly on the wall here, but @JakeWharton 's suggestion sounds reasonable to this observer.

**Comment by JakeWharton** [comment]
Had a bunch of nits, but we'll need to do a pass over everything anyway so whevs. Forward progress is more important.

File: okio/jvm/src/main/java/okio/Buffer.kt (hunk 3 above)
Comment (swankjesse) [reply okio/jvm/src/main/java/okio/Buffer.kt]: I ran into a wall trying to get the `size()` method on Options to work. I think we quietly release Okio 1.14.1 which just adds a bunch of redundant final modifiers.

https://github.com/square/okio/pull/417/files

**Pull request merged**